rustup-init.ps1

```diff
@@ -0,0 +1,142 @@
+        --default-toolchain <default-toolchain>    Choose a default toolchain to install
+'@
+}
+
+function Invoke-Main([AllowEmptyCollection()] [string[]]$Argv = @()) {
+    foreach ($arg in $Argv) {
+        if ("$arg" -eq "-h" -or "$arg" -eq "--help") {
+            Write-Usage
+            return
+        }
+    }
+
+    Install-Rustup $Argv
+}
+
+function Get-Architecture {
+    if (($IsWindows -eq $null) -or $IsWindows) {
+        $ostype = "pc-windows-msvc"
+    } elseif ($IsMacOS) {
+        $ostype = "apple-darwin"
+    } elseif ($IsLinux) {
+        $ostype = "unknown-linux-gnu"
+    } else {
+        Write-Failure "unrecognized OS type"
+    }
+
+    if ([System.IntPtr]::Size -eq 8) {
+        $cputype = "x86_64"
+    } elseif ([System.IntPtr]::Size -eq 4) {
+        $cputype = "i686"
+    } else {
+        Write-Failure "unknown CPU type"
+    }
+
+    $arch = "${cputype}-${ostype}"
+    Write-Verbose "Detected arch '$arch'"
+    $arch
+}
+
+# If RUSTUP_UPDATE_ROOT is unset, default it.
+function Get-RustupUpdateRoot {
+    $EnvVar = [System.Environment]::GetEnvironmentVariable("RUSTUP_UPDATE_ROOT")
+    if (($EnvVar -ne $null) -and (Test-Path -Path $EnvVar)) {
+        $EnvVar
+    } else {
+        "https://static.rust-lang.org/rustup"
+    }
+}
+
+function Install-Rustup([AllowEmptyCollection()] [string[]]$Argv) {
+    $arch = Get-Architecture
+    switch -Wildcard ($arch) {
+        "*windows*" { $ext = ".exe"; break }
+        default { $ext = "" }
+    }
+
+    $url = "$(Get-RustupUpdateRoot)/dist/$arch/rustup-init$ext"
+
+    $dir = New-TemporaryDirectory
+    $file = Join-Path $dir "rustup-init$ext"
+
+    try {
+        if (($IsWindows -eq $null) -or $IsWindows) { Set-HighestEncryption }
+
+        Write-Verbose "Downloading $url to $file"
+        (New-Object System.Net.WebClient).DownloadFile($url, $file)
+        if ($IsMacOS -or $IsLinux) {
+          Write-Verbose "Setting execute permission on $file"
+          chmod u+x "$file"
+        }
+
+        Write-Verbose "Calling $file $Argv"
+        & "$file" @Argv
+    } finally {
+        Write-Verbose "Cleaning $dir"
+        Remove-Item "$dir" -Force -Recurse
+    }
+}
+
+function New-TemporaryDirectory {
+    $parent = [System.IO.Path]::GetTempPath()
+    [string]$name = [System.Guid]::NewGuid()
+    New-Item -ItemType Directory -Path (Join-Path $parent $name)
+}
+
+function Set-HighestEncryption {
+    # Implementation adapted from the Chocolatey installer script,
+    # see: https://chocolatey.org/install.ps1
+
+    # Attempt to set highest encryption available for SecurityProtocol.
+    # PowerShell will not set this by default (until maybe .NET 4.6.x). This
+    # will typically produce a message for PowerShell v2 (just an info message
+    # though)
+    $OldSPM = [System.Net.ServicePointManager]::SecurityProtocol
+    try {
+        # Set TLS 1.2 (3072). Use integers because the enumeration values for
```

Comment:
I think its worth a note in here about why you are not setting TLS 1.3. (presumably because static.r-l.org isn't TLS1.3 enabled yet, but whatever the reason, please do capture it.

Reply:
I added a bit more explanation about why we're not yet setting TLS v1.3--short reason being I don't believe that it's supported on https://static.rust-lang.org currently. The comment calls this out and provides the enum value for TLS v1.3 so we can add that when appropriate.

Additionally after looking at this behavior, it seemed appropriate to set this for macOS and Linux. I tested these platforms out and they all behaved as expected.